lib/dao.js

```diff
@@ -624,7 +622,208 @@ DataAccessObject.upsert = function(data, options, cb) {
   }
   return cb.promise;
 };
+/**
+ * Performs upsert based on the instance that matches the where clause
+ . This is a PUT operation and based on
+ * non-primary key. The filter takes one key value pair and updates single matching instance.*
```

Comment:
@Amir-61  cc @bajtos .. This is as per the requirement we have .  Idea is to have an operation which will update single instance based on the non-primary key. These 3 conditions is what comes as part of our requirement :  1) Do a retrieve operation based on a non-primary key and update, if there is a single matching instance, throw error, if multiple instances are fetched and create if there are no instances found. For primary key, we have updateOrCreate operation already and updateAll performs batch update based on the filter. This is a different version which we wanted. And in future, if there is a need to support multiple key value pair , we will extend it for sure.